### ISSUE TITLE
TV Shows and Seasons focus issues.

### ISSUE THREAD

**Issue by WolfganP** [NineWorlds/serenity-android#367]
@kingargyle the newest release was a success in providing a working app as per https://github.com/NineWorlds/serenity-android/issues/365 (M3 logs pasted below for your reference).

The only issue I observed is in Grid View, while navigating the different items with the remote, the background image is changing but the white border indicating the item focus remains fixed in the first item.
Thanks for the prompt M3 build!

```
E/InstalledAppDetails( 1150): Exception when retrieving package:us.nineworlds.serenity
E/InstalledAppDetails( 1150): android.content.pm.PackageManager$NameNotFoundException: us.nineworlds.serenity
E/WorkSourceUtil(  951): Could not find package: us.nineworlds.serenity
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.GlideClearCacheJob@41f817f0
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.GlideClearCacheJob.onRun(GlideClearCacheJob.java:30)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.GlideClearCacheJob@41f817f0
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.GlideClearCacheJob.onRun(GlideClearCacheJob.java:30)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.GlideClearCacheJob@41f817f0
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.GlideClearCacheJob.onRun(GlideClearCacheJob.java:30)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.GlideClearCacheJob@41f817f0
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.GlideClearCacheJob.onRun(GlideClearCacheJob.java:30)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.GlideClearCacheJob@41f817f0
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.GlideClearCacheJob.onRun(GlideClearCacheJob.java:30)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.GlideClearCacheJob@41f817f0
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.GlideClearCacheJob.onRun(GlideClearCacheJob.java:30)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.GlideClearCacheJob@41f817f0
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.GlideClearCacheJob.onRun(GlideClearCacheJob.java:30)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.GlideClearCacheJob@41f817f0
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.GlideClearCacheJob.onRun(GlideClearCacheJob.java:30)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.GlideClearCacheJob@41f817f0
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.GlideClearCacheJob.onRun(GlideClearCacheJob.java:30)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.GlideClearCacheJob@41f817f0
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.GlideClearCacheJob.onRun(GlideClearCacheJob.java:30)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.GlideClearCacheJob@41f817f0
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.GlideClearCacheJob.onRun(GlideClearCacheJob.java:30)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.GlideClearCacheJob@41f817f0
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.GlideClearCacheJob.onRun(GlideClearCacheJob.java:30)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.GlideClearCacheJob@41f817f0
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.GlideClearCacheJob.onRun(GlideClearCacheJob.java:30)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.GlideClearCacheJob@41f817f0
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.GlideClearCacheJob.onRun(GlideClearCacheJob.java:30)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.GlideClearCacheJob@41f817f0
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.GlideClearCacheJob.onRun(GlideClearCacheJob.java:30)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.GlideClearCacheJob@41f817f0
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.GlideClearCacheJob.onRun(GlideClearCacheJob.java:30)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.GlideClearCacheJob@41f817f0
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.GlideClearCacheJob.onRun(GlideClearCacheJob.java:30)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.GlideClearCacheJob@41f817f0
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.GlideClearCacheJob.onRun(GlideClearCacheJob.java:30)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.GlideClearCacheJob@41f817f0
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.GlideClearCacheJob.onRun(GlideClearCacheJob.java:30)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.GlideClearCacheJob@41f817f0
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.GlideClearCacheJob.onRun(GlideClearCacheJob.java:30)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.MainMenuRetrievalJob@4209c750
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.MainMenuRetrievalJob.onRun(MainMenuRetrievalJob.java:24)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.MainMenuRetrievalJob@4209c750
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.MainMenuRetrievalJob.onRun(MainMenuRetrievalJob.java:24)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.MainMenuRetrievalJob@4209c750
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.MainMenuRetrievalJob.onRun(MainMenuRetrievalJob.java:24)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.MainMenuRetrievalJob@4209c750
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.MainMenuRetrievalJob.onRun(MainMenuRetrievalJob.java:24)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.MainMenuRetrievalJob@4209c750
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.MainMenuRetrievalJob.onRun(MainMenuRetrievalJob.java:24)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.MainMenuRetrievalJob@4209c750
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.MainMenuRetrievalJob.onRun(MainMenuRetrievalJob.java:24)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.MainMenuRetrievalJob@4209c750
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.MainMenuRetrievalJob.onRun(MainMenuRetrievalJob.java:24)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.MainMenuRetrievalJob@4209c750
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.MainMenuRetrievalJob.onRun(MainMenuRetrievalJob.java:24)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.MainMenuRetrievalJob@4209c750
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.MainMenuRetrievalJob.onRun(MainMenuRetrievalJob.java:24)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.MainMenuRetrievalJob@4209c750
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.MainMenuRetrievalJob.onRun(MainMenuRetrievalJob.java:24)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.MainMenuRetrievalJob@4209c750
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.MainMenuRetrievalJob.onRun(MainMenuRetrievalJob.java:24)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.MainMenuRetrievalJob@4209c750
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.MainMenuRetrievalJob.onRun(MainMenuRetrievalJob.java:24)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.MainMenuRetrievalJob@4209c750
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.MainMenuRetrievalJob.onRun(MainMenuRetrievalJob.java:24)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.MainMenuRetrievalJob@4209c750
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.MainMenuRetrievalJob.onRun(MainMenuRetrievalJob.java:24)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.MainMenuRetrievalJob@4209c750
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.MainMenuRetrievalJob.onRun(MainMenuRetrievalJob.java:24)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.MainMenuRetrievalJob@4209c750
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.MainMenuRetrievalJob.onRun(MainMenuRetrievalJob.java:24)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.MainMenuRetrievalJob@4209c750
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.MainMenuRetrievalJob.onRun(MainMenuRetrievalJob.java:24)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.MainMenuRetrievalJob@4209c750
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.MainMenuRetrievalJob.onRun(MainMenuRetrievalJob.java:24)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.MainMenuRetrievalJob@4209c750
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.MainMenuRetrievalJob.onRun(MainMenuRetrievalJob.java:24)
E/JobManager( 3681): error while executing job us.nineworlds.serenity.jobs.MainMenuRetrievalJob@4209c750
E/JobManager( 3681): 	at us.nineworlds.serenity.jobs.MainMenuRetrievalJob.onRun(MainMenuRetrievalJob.java:24)
E/ActivityThread( 3681): Performing pause of activity that is not resumed: {us.nineworlds.serenity/us.nineworlds.serenity.MainActivity}
E/ActivityThread( 3681): java.lang.RuntimeException: Performing pause of activity that is not resumed: {us.nineworlds.serenity/us.nineworlds.serenity.MainActivity}
```

**Comment by kingargyle** [comment]
Yeah there are a couple other minor bugs around Grid views, when switching back and forth between Detail and Grid on movies.   Does this happen only with Movies, or does it happen with TV Shows as well?

**Comment by WolfganP** [comment]
Focus navigation with remote control is kinda weird, as sometimes is not clear where the focus is and what the next action will be.
As far as I recall it happened in all categories, including while navigating in Detail view (when the focus is supposedly at the mid level, items list).
BTW, for using Serenity in an Android box, do you suggest to use the standard or the TV UI option? (as per the app settings)

**Comment by kingargyle** [comment]
TV UI settings if you are using an Remote control.   I'll try and fire up an Emulator for a 4.4 device and see if I can replicate the problems.  My main focus is for Android TV and Fire TV devices right now.

**Comment by WolfganP** [comment]
Thx @kingargyle. I found that I can't navigate and start playing any TV show with seasons, as the focus never seems to reach an individual episode, even using a mouse the individual episodes can't be triggered. I'll produce specific logs for that issue later today.
Thanks again!

**Comment by kingargyle** [comment]
Okay, one thing to check if you are willing to mess around with the code and layouts, we might have to be more aggressive on setting  android:duplicateParentState="true" on some of the elements.   Specifically if they are wrapped in the DPadAwareRecyclerview.

In the past when I was experiencing a similar issue on Android 5.0 or higher, it was because the state not being passed correctly.   As for the mouse, yeah, I'm not surprised.  The current version of the app doesn't implement the standard onClick events, it uses the ones provided by the DPadAwareRecyclerView.    I may look at swapping that version out for some more of the Android Leanback components which should help address some of the focus and clicking issues.

**Comment by kingargyle** [comment]
@WolfganP I've opened issue #368 to address the focus and navigation issues you are exeriencing.  I think many of those issues can be fixed by migrating over to the Android Leanback libraries HorizontalGridView implementation.

https://developer.android.com/reference/android/support/v17/leanback/widget/HorizontalGridView.html

**Comment by WolfganP** [comment]
> Okay, one thing to check if you are willing to mess around with the code and layouts, we might have to be more aggressive on setting android:duplicateParentState="true" on some of the elements. Specifically if they are wrapped in the DPadAwareRecyclerview.

Sounds good. Let me know what do you want me to check and I'll do it. Besides, if you want to produce a version with extended log info that helps troubleshooting just point me to it and I'll install & test it.

**Comment by kingargyle** [comment]
One thing I might be able to do is make sure that the focus on the existing recyclerViews has focusAfterDecendants set, to help make sure that the child views are getting focus.   I'll try that first this week, and can hopefully get some thing for you to test in a few days.

**Comment by kingargyle** [comment]
@WolfganP did some small tweaks trying to improve the focusability.  I tested on a Nexus 7 emulator running KitKit, and wasn't able to reproduce the loss of focus on the TV Shows or Movies screen.  Regardless I added the descendats focusability = afterDescendats to all the Recycler Views, so hopefully this improves it some.

If not, I'll have to up the time line in migrating off DPadAware Recycler and move to the Leanback ListRow fragment.

**Comment by kingargyle** [comment]
https://github.com/NineWorlds/serenity-android/releases/tag/v2.0.0-M4-wolfgang

**Comment by WolfganP** [comment]
@kingargyle I tried the M4 version but it didn't make any change in behavior of the focus, and weirdly, it didn't produce any logs to upload compared to the previous versions.
Just to be accurate on the issue description, this only happen in TV Shows & seasons (Movies and Music videos work OK thou as they are shown in a planar view). I don't know if this is being experienced by other users as well.

Example showing Seasons from a TV series (no focus in any of the individual episodes in mid section, also note the white border indicating the focus to season 8 when the display shows season 6, seems like the graphic element is off sync the logical focus):

![serenity-seasons](https://user-images.githubusercontent.com/2248211/30331257-b0ef0388-97ad-11e7-8b25-9408f7661efc.jpg)

And entering an specific season (in this case I can't navigate any element in the lower section, nor a description is pulled for the specific episode):

![serenity-seasonx](https://user-images.githubusercontent.com/2248211/30331294-c530d236-97ad-11e7-8455-53fe6680d4b4.jpg)

Thanks!

**Comment by kingargyle** [comment]
Okay, I'll fire up my Android Emulator again and specifically test against.  I unfortunately don't experience any of these issues on the Fire TV or MiBox Android TV device I use for most of my development.  I suspect when I migrate to the Leanback library for the same functionality a lot of the issues will disappear.

**Comment by WolfganP** [comment]
Thx for the prompt feedback. Whatever you want me to test, just let me know!

**Comment by kingargyle** [comment]
Probably won't be until next week sometime. But keep an eye on this and the other issues for 2.0, and I'll post when 2.0.0-M4 is ready.

**Comment by kingargyle** [comment]
Issue being addressed in #378 .  Closing this out for now.
